### PR TITLE
Partition leading zeros in baseX in one step using regular expressions

### PR DESCRIPTION
**Pull request by steveluscher**
# Test Plan

No change in benchmarks.

```shell
cd packages/codecs-strings/
pnpm benchmark
```

### CODE REVIEW

**Comment by steveluscher** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2141?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #2139** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2139?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #2140** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2140?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #2141** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2141?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #2142** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2142?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2141?utm_source=stack-comment).

File: packages/codecs-strings/src/baseX.ts [hunk 1]
```diff
@@ -92,9 +92,12 @@ export const getBaseXDecoder = (alphabet: string): VariableSizeDecoder<string> =
...
     combineCodec(getBaseXEncoder(alphabet), getBaseXDecoder(alphabet));
 
-function partitionLeadingZeroes(value: string, zeroCharacter: string): [string, string] {
-    const leadingZeroIndex = [...value].findIndex(c => c !== zeroCharacter);
-    return leadingZeroIndex === -1 ? [value, ''] : [value.slice(0, leadingZeroIndex), value.slice(leadingZeroIndex)];
+function partitionLeadingZeroes(
+    value: string,
+    zeroCharacter: string,
+): [leadingZeros: string, tailChars: string | undefined] {
+    const [leadingZeros, tailChars] = value.split(new RegExp(`((?!${zeroCharacter}).*)`));
```
Comment (steveluscher) [inline packages/codecs-strings/src/baseX.ts:99]: This regular expression says ‘split on the first non-zeroCharacter and onward.’ This naturally gives us a partition, which we can output directly.

Comment (lorisleiva) [reply packages/codecs-strings/src/baseX.ts:99]: Nice! Should we just return that line directly instead of allocating a new array?

**Review by lorisleiva — APPROVED** [review]

File: packages/codecs-strings/src/baseX.ts (hunk 1 above)
Comment (steveluscher) [reply packages/codecs-strings/src/baseX.ts:99]: Annoyingly, the return is `[leading, tailing, emptyString]`.

**Comment by steveluscher** [comment]
### Merge activity

* **Feb 20, 11:30 AM PST**: @steveluscher started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2141).

**Pull request merged**